### ISSUE TITLE
Exposed slow

### ISSUE THREAD

**Issue by MGlolenstine** [JetBrains/Exposed#776]
I'm trying to connect to the H2 database on an embedded device.

The connection using 
```
Database.connect("jdbc:h2:~/testdb;DB_CLOSE_DELAY=-1", driver = "org.h2.Driver", user = "root", password = "")
```
takes about 7s to create/open a database, but the connection using another library, for example [**kotliquery** ](https://search.maven.org/artifact/com.github.seratch/kotliquery/1.3.1/jar)
```
val create = measureTimeMillis {
            sess = sessionOf("jdbc:h2:~/testdb;DB_CLOSE_DELAY=-1", "root", "")
            sess.execute(queryOf("CREATE TABLE IF NOT EXISTS RandomText(ID INT PRIMARY KEY auto_increment, text VARCHAR(15));"))
        }
```
takes about 2s.

Any reason for that overhead and is there any way to speed it up?

We're running the program and the database on an embedded device called Colibri IMx6, and the speedup would greatly help us.

**Comment by Tapac** [comment]
Could you share code or tell what exactly runs so long? 
Do you call `SchemaUtils.create(table)` or `SchemaUtils.createMissingTablesAndColumns(table)` ?

**Comment by MGlolenstine** [comment]
@Tapac I do call `SchemaUtils.create(table)`, but it never reset the database, so I thought that it was automatically detecting that the database was there and it didn't create it again. I'll test the other one out tomorrow!

**Comment by MGlolenstine** [comment]
@Tapac I'm back with more information.
I just ran this piece of code, to test out the performance with `SchemaUtils#createMissingTablesAndColumns()`, but I think that the performance is even lower.
```
Database.connect("jdbc:h2:~/testdb;DB_CLOSE_DELAY=-1", driver = "org.h2.Driver", user = "root", password = "")
transaction {
    val create = measureTimeMillis { SchemaUtils.createMissingTablesAndColumns(RandomText) }
    println("h2 database creation took $create ms")
    val fillingDb = measureTimeMillis {
        fillDb()
    }
}```
This one responds with `h2 database creation took 8193 ms`
And the normal `SchemaUtils#create()`
```
Database.connect("jdbc:h2:~/testdb;DB_CLOSE_DELAY=-1", driver = "org.h2.Driver", user = "root", password = "")
transaction {
    val create = measureTimeMillis { SchemaUtils.create(RandomText) }
    println("h2 database creation took $create ms")
    val fillingDb = measureTimeMillis {
        fillDb()
    }
}
```
responds with `h2 database creation took 7206 ms`

Keep in mind, that this is not on the first run, but on the second consecutive run, so the database is already created, so this is just the "Connection" to it.

I've noticed that I haven't posted my database yet, so here it is:
```
object RandomText : Table() {
    val id = integer("id").autoIncrement()
    val text = varchar("text", length = 15)
    override val primaryKey = PrimaryKey(id)
}
```

**Comment by Tapac** [comment]
How many tables do you have in your testdb?

**Comment by MGlolenstine** [comment]
There's only one and I sent it in the post before.

**Comment by Tapac** [comment]
Could you check the same code on a new empty database like `test2,db`?
And also, please check the value of `TransactionManager.current().db.dialect.allTablesNames.size`?

**Comment by MGlolenstine** [comment]
Before reporting the status, I've deleted the databases and ran it once, then restarted the program (so it should only contain that one table, keep in mind, that the embedded platform runs on ARMv6 and 256MB of RAM). First run usually took about 2s longer.
I can get more benchmarks and a runnable example on Monday.

**Comment by Tapac** [comment]
That would be great.

**Comment by MGlolenstine** [comment]
Ok, here I have some benchmarks @Tapac:

Benchmark type | time
--- | ---
`SchemaUtils#create` with no database | 16750 ms
`SchemaUtils#create` with an existing database | 11448 ms
`SchemaUtils#createMissingTablesAndColumns` with no database | 17660 ms
`SchemaUtils#createMissingTablesAndColumns` with an existing database | 12141 ms

**Comment by Tapac** [comment]
@MGlolenstine , could you, please, add "plain jdbc" or with using another lib (like kotliquery) to compare with?

**Comment by MGlolenstine** [comment]
Sure thing! I'm on it!

**Comment by MGlolenstine** [comment]
@Tapac Anything new?

**Comment by Tapac** [comment]
@MGlolenstine , sorry I don't have much time on this weekday, will try to look at weekends.

**Comment by MGlolenstine** [comment]
@Tapac It's of no bother. Look at it when you have time :)
It's just that it's been quiet and I didn't know what's happening :)

**Comment by Tapac** [comment]
After some investigation, I was able to localize the main problem with the first long initialization - the cause is kotlin reflection cache long loading on the first access to `RandomTable` (see similar problem on [reddit](https://www.reddit.com/r/Kotlin/comments/e7h6f2/kotlinreflect_startup_time/). And there is not so much what I can do without complicating the code.

Another overhead is on object creation as Exposed trying to fetch generated data after insert. 

I will profile some common cases a bit more but can't guaranty that I can radically improve the performance. I was able to achieve similar performance on inserts by using `batchInsert` or executing the plain sql, but it's not very fair:D
